zrml/prediction-markets/src/lib.rs

```diff
@@ -84,13 +84,92 @@ mod pallet {
+                        );
+                        log::warn!("{}", warning);
+                        debug_assert!(false, "{}", warning);
+                        return Ok(());
+                    }
+                    CurrencyOf::<T>::unreserve_named(&Self::reserve_id(), &bond.who, bond.value);
+                    market.bonds.$bond_type = Some(Bond { is_settled: true, ..bond.clone() });
+                    Ok(())
+                })
+            }
+        };
+    }
+
+    macro_rules! impl_slash_bond {
+        ($fn_name:ident, $bond_type:ident) => {
+            /// Settle the $bond_type bond by slashing it and return the resulting imbalance.
+            ///
+            /// This function **should** only be called if the bond is not yet settled, and calling
+            /// it if the bond is settled is most likely a logic error. If the bond is already
+            /// settled, storage is not changed, a warning is raised and a zero imbalance is
+            /// returned.
+            fn $fn_name(
+                market_id: &MarketIdOf<T>,
+                market: &MarketOf<T>,
+            ) -> Result<NegativeImbalanceOf<T>, DispatchError> {
+                let bond = market.bonds.$bond_type.as_ref().ok_or(Error::<T>::MissingBond)?;
+                // Trying to settle a bond multiple times is always a logic error, not a runtime
+                // error, so we log a warning instead of raising an error.
+                if bond.is_settled {
+                    let warning = format!(
+                        "Attempting to settle the {} bond of market {:?} multiple times",
+                        stringify!($bond_type),
+                        market_id,
+                    );
+                    log::warn!("{}", warning);
+                    debug_assert!(false, "{}", warning);
+                    return Ok(NegativeImbalanceOf::<T>::zero());
+                }
+                let (imbalance, excess) = CurrencyOf::<T>::slash_reserved_named(
+                    &Self::reserve_id(),
+                    &bond.who,
+                    bond.value,
+                );
+                // If there's excess, there's nothing we can do, so we don't count this as error
+                // and log a warning instead.
+                if excess != BalanceOf::<T>::zero() {
+                    let warning = format!(
+                        "Failed to settle the {} bond of market {:?}",
+                        stringify!($bond_type),
+                        market_id,
+                    );
+                    log::warn!("{}", warning);
+                    debug_assert!(false, "{}", warning);
+                }
+                <zrml_market_commons::Pallet<T>>::mutate_market(market_id, |m| {
+                    m.bonds.$bond_type = Some(Bond { is_settled: true, ..bond.clone() });
+                    Ok(())
+                })?;
+                Ok(imbalance)
+            }
+        };
+    }
```

Comment:
https://docs.substrate.io/build/runtime-storage/

> runtime storage involves an underlying key-value database and in-memory storage overlay abstractions that keep track of keys and state changes until the values are committed to the underlying database. By default, functions in the runtime write changes to a single in-memory transactional storage layer before committing them to the main storage overlay. If an error prevents the transaction from being completed, the changes in the transactional storage layer are discarded instead of being passed on to the main storage overlay and state in the underlying database remains unchanged

Reply:
I like both ways, both ways have their pros and cons.

> It has the disadvantage that the market potentially has to be read and written multiple times from storage, although the market is already present in memory.

Only written in the transactional storage layer (Storage Overlay), so no expensive database operation.

Reply:
I don't understand why a mutate is necessary here. I would have imagined this:
```rust
let mut market = <zrml_market_commons::Pallet<T>>::market(market_id)?;
let bond = market.bonds.$bond_type.as_ref().ok_or(Error::<T>::MissingBond)?;

if bond.is_settled {
    let warning = format!(
        "Attempting to settle the {} bond of market {:?} multiple times",
        stringify!($bond_type),
        market_id,
    );
    log::warn!("{}", warning);
    debug_assert!(false, "{}", warning);
    return Ok(());
}

CurrencyOf::<T>::unreserve_named(&Self::reserve_id(), &bond.who, bond.value);
market.bonds.$bond_type = Some(Bond { is_settled: true, ..bond.clone() });
<zrml_market_commons::Pallet<T>>::insert_market(market_id, market)
```

Seems like `MarketCommonsPalletApi` is incomplete here. We can fetch a market but we cannot insert a single market, we are always forced to mutate.
Alternatively we could directly access the storage `<<zrml_market_commons::Markets<T>>::insert(market_id, market)`, but I am hugely in favor of providing a proper API and using it instead.

Reply:
Yes, we should definitely have an API call for this. I guess `mutate` has the advantage of verifying that the market exists.